examples/cms/src/FieldMapping.tsx

```diff
@@ -0,0 +1,233 @@
+                strokeLinecap="round"
+                strokeLinejoin="round"
+            />
+        </svg>
+    )
+}
+
+interface FieldMappingRowProps {
+    field: ManagedCollectionField
+    isIgnored: boolean
+    onFieldToggle: (fieldId: string) => void
+    onFieldNameChange: (fieldId: string, name: string) => void
+}
+
+const FieldMappingRow = memo(({ field, isIgnored, onFieldToggle, onFieldNameChange }: FieldMappingRowProps) => {
+    const [hasCustomName, setHasCustomName] = useState(field.id !== field.name)
+
+    return (
+        <>
+            <div
+                className="source-field"
+                aria-disabled={isIgnored}
+                onClick={() => onFieldToggle(field.id)}
+                role="button"
+            >
+                <input
+                    type="checkbox"
+                    checked={!isIgnored}
+                    onChange={event => {
+                        event.stopPropagation()
+                    }}
+                />
+                <span>{field.id}</span>
+            </div>
+            <ChevronIcon />
+            <input
+                type="text"
+                style={{
+                    width: "100%",
+                    opacity: isIgnored ? 0.5 : 1,
+                }}
+                disabled={isIgnored}
+                placeholder={field.id}
+                value={hasCustomName ? field.name : ""}
+                onChange={event => {
+                    const value = event.target.value
+                    if (!value.trim()) {
+                        setHasCustomName(false)
+                        onFieldNameChange(field.id, field.id)
+                    } else {
+                        setHasCustomName(true)
+                        onFieldNameChange(field.id, value.trimStart())
+                    }
+                }}
+            />
+        </>
+    )
+})
+
+interface FieldMappingProps {
+    collection: ManagedCollection
+    dataSource: DataSource
+    initialSlugFieldId: string | null
+}
+
+type FieldMappingStatus = "mapping-fields" | "loading-fields" | "syncing-collection"
+
+export function FieldMapping({ collection, dataSource, initialSlugFieldId }: FieldMappingProps) {
+    const [status, setStatus] = useState<FieldMappingStatus>("loading-fields")
+    const isSyncing = status === "syncing-collection"
+    const isLoadingFields = status === "loading-fields"
+
+    const sourceFields = useMemo(() => computeFieldsFromDataSource(dataSource), [dataSource])
+    const possibleSlugFields = useMemo(() => sourceFields.filter(field => field.type === "string"), [sourceFields])
+
+    const [selectedSlugField, setSelectedSlugField] = useState<ManagedCollectionField | null>(
+        possibleSlugFields.find(field => field.id === initialSlugFieldId) ?? possibleSlugFields[0] ?? null
+    )
+
+    const [fields, setFields] = useState<ManagedCollectionField[]>([])
+    const [ignoredFieldIds, setIgnoredFieldIds] = useState<Set<string>>(new Set())
```

Comment:
I thin it would be good to mark these as readonly. And I wonder if the defaults should be lifted outside of rendering, so they don't have to be recreated on every render.

```ts
// Outside of rendering
type ManagedCollectionFields = readonly ManagedCollectionField[]
const initialManagedCollectionFields: ManagedCollectionFields = []

type FieldId = string
type FieldIds = ReadonlySet<FieldId>
const initialFieldIds: FieldIds = new Set()

// Within the component
const [fields, setFields] = useState<ManagedCollectionFields>(initialManagedCollectionFields)
const [ignoredFieldIds, setIgnoredFieldIds] = useState<FieldIds>(initialFieldIds)
```